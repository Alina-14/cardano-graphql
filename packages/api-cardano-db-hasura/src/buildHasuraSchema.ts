import { fetch } from 'cross-fetch'
import { introspectSchema, wrapSchema } from '@graphql-tools/wrap'
import { DocumentNode, print } from 'graphql'

export async function buildHasuraSchema (hasuraUri: string) {
  const executor = async ({ document, variables }: { document: DocumentNode, variables?: Object }) => {
    const query = print(document)
    try {
      const fetchResult = await fetch(`${hasuraUri}/v1/graphql`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Hasura-Role': 'cardano-graphql'
        },
        body: JSON.stringify({ query, variables })
      })
      return fetchResult.json()
    } catch (error) {
      console.error(error)
      throw error
    }
  }

  return wrapSchema({
    schema: await introspectSchema(executor),
    executor
  })
}
