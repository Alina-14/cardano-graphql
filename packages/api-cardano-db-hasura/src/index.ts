export * from './Db'
export * from './executableSchema'
export * from './graphql_types'
