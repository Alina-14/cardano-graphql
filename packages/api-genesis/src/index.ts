export * from './Config'
export * from './executableSchema'
export * from './graphql_types'
