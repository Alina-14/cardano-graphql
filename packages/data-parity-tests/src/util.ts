import fetch from 'node-fetch'

export async function getDataFromAPI (path: string) {
  const response = await fetch(`http://localhost:8201/${path}`)
  return response.json()
}
