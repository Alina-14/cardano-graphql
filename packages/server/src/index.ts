import { getConfig } from './config'
import { CompleteApiServer } from './CompleteApiServer'

export * from './config'

(async function () {
  try {
    const server = await CompleteApiServer(await getConfig())
    await server.init()
    await server.start()
  } catch (error) {
    console.error(error.message)
    process.exit(1)
  }
})()
