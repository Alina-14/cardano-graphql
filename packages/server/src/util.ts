import { Application } from 'express'
import http from 'http'

export function listenPromise (app: Application, config: { port: number }): Promise<http.Server> {
  return new Promise(function (resolve, reject) {
    const server: http.Server = app.listen(config.port, () => resolve(server))
    server.on('error', reject)
  })
}
