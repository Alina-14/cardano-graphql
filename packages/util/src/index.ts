import { onFailedAttemptFor } from './onFailedAttemptFor'
import { loadQueryNode } from './queryNodeLoading'

export default {
  onFailedAttemptFor,
  loadQueryNode
}
