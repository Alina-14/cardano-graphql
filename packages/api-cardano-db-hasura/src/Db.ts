import { ApolloClient, gql, InMemoryCache, NormalizedCacheObject } from 'apollo-boost'
import { createHttpLink } from 'apollo-link-http'
import fetch from 'cross-fetch'
import utc from 'dayjs/plugin/utc'
import dayjs from 'dayjs'
import { exec } from 'child_process'
import { clearIntervalAsync, setIntervalAsync, SetIntervalAsyncTimer } from 'set-interval-async/dynamic'
import path from 'path'
import pRetry from 'p-retry'
import util from '@cardano-graphql/util'

dayjs.extend(utc)

export class Db {
  hasuraClient: ApolloClient<NormalizedCacheObject>
  hasuraUri: string
  monitorTimer: SetIntervalAsyncTimer

  constructor (hasuraUri: string) {
    this.hasuraUri = hasuraUri
    this.hasuraClient = new ApolloClient({
      cache: new InMemoryCache({
        addTypename: false
      }),
      defaultOptions: {
        query: {
          fetchPolicy: 'network-only'
        }
      },
      link: createHttpLink({
        uri: `${this.hasuraUri}/v1/graphql`,
        fetch,
        headers: {
          'X-Hasura-Role': 'cardano-graphql'
        }
      })
    })
  }

  public async init (): Promise<void> {
    // Todo: optimal solution dependent on https://github.com/input-output-hk/cardano-db-sync/issues/182
    await this.applySchemaAndMetadata()
    this.monitorDbState()
  }

  public async shutdown () {
    await clearIntervalAsync(this.monitorTimer)
  }

  private monitorDbState () {
    this.monitorTimer = setIntervalAsync(
      async () => {
        try {
          await this.getMeta()
        } catch (error) {
          if (error.message === 'GraphQL error: field "cardano" not found in type: \'query_root\'' || error.message === 'GraphQL error: database query error') {
            console.warn('Re-applying PostgreSQL migrations and Hasura metadata')
            await this.applySchemaAndMetadata()
          } else {
            console.error(error)
          }
        }
      },
      10000
    )
  }

  async getMeta () {
    const result = await this.hasuraClient.query({
      query: gql`query {
          cardano {
              tip {
                  forgedAt
              }
          }}`
    })
    const { tip } = result.data?.cardano[0]
    const currentUtc = dayjs().utc()
    const tipUtc = dayjs.utc(tip.forgedAt)
    return {
      initialized: tipUtc.isAfter(currentUtc.subtract(120, 'second')),
      syncPercentage: (tipUtc.valueOf() / currentUtc.valueOf()) * 100
    }
  }

  public async applySchemaAndMetadata () {
    await pRetry(async () => {
      await this.hasuraCli('migrate apply --down all')
      await this.hasuraCli('migrate apply --up all')
      await this.hasuraCli('metadata clear')
      await this.hasuraCli('metadata apply')
    }, {
      factor: 1.75,
      retries: 9,
      onFailedAttempt: util.onFailedAttemptFor('Applying PostgreSQL schema and Hasura metadata')
    })
  }

  private hasuraCli (command: string) {
    return new Promise((resolve, reject) => {
      exec(
        `hasura --skip-update-check --project ${path.resolve(__dirname, '..', 'hasura', 'project')} --endpoint ${this.hasuraUri} ${command}`,
        (error, stdout) => {
          if (error) {
            reject(error)
          }
          console.log(stdout)
          resolve()
        }
      )
    })
  }
}
