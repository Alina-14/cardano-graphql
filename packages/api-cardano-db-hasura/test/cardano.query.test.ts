import path from 'path'

import { DocumentNode } from 'graphql'
import util from '@cardano-graphql/util'
import utilDev, { TestClient } from '@cardano-graphql/util-dev'
import { buildSchema } from '../src'

function loadQueryNode (name: string): Promise<DocumentNode> {
  return util.loadQueryNode(path.resolve(__dirname, '..', 'src', 'example_queries', 'cardano'), name)
}

describe('cardano', () => {
  let client: TestClient
  beforeAll(async () => {
    if (process.env.TEST_MODE === 'e2e') {
      client = await utilDev.createE2EClient()
    } else {
      const schema = await buildSchema('http://localhost:8090')
      client = await utilDev.createIntegrationClient(schema)
    }
  }, 60000)

  it('Returns static information about the network', async () => {
    const result = await client.query({
      query: await loadQueryNode('static')
    })
    expect(result.data).toMatchSnapshot()
  })
  it('Returns dynamic information about the network', async () => {
    const result = await client.query({
      query: await loadQueryNode('dynamic')
    })
    expect(result.data.cardano.blockHeight).toBeGreaterThan(3994551)
    expect(result.data.cardano.currentEpoch.number).toBeGreaterThan(184)
  })
})
