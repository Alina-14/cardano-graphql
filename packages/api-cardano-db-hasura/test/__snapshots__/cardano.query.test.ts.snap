// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cardano Returns static information about the network 1`] = `
Object {
  "cardano": Object {
    "networkName": "mainnet",
    "slotsPerEpoch": 21600,
    "startTime": "2017-09-23T21:44:51.000Z",
  },
}
`;
